# Link github: 
https://github.com/nightfury1103/P4DS-CQ19-FinalProject

# Import

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Thu thập dữ liệu

## Dữ liệu của bạn là về chủ đề gì 

- Tập dữ liệu đưa ra các mô tả ngắn gọn về các cơn sóng thần xảy ra từ năm 1980 đến năm 2021
- Có tổng cộng 26 cột với các thông tin từ ngày tháng đến các thống kê về mức độ thiệt hại

## Nguồn dữ liệu - Quyền sử dụng

- Dữ liệu sử dụng của notebook được public trên trang web **kaggle** với đường dẫn sau: [link](https://www.kaggle.com/ramjasmaurya/historical-data-of-tsunamis18002021).
- Về vấn đề quyền sử dụng: Tác giả đề cập đến vấn đề quyền sử dụng dựa theo [CC0 1.0 Universal (CC0 1.0) Public Domain Dedication](https://creativecommons.org/publicdomain/zero/1.0/)

## Người ta thu thập dữ liệu như thế nào?

- Dữ liệu được thu thập từ trang https://www.ngdc.noaa.gov/hazard/tsu_db.shtml với các cơ sở dữ liệu **Tsunami Runup** và **Tsunami Source Event**

## Mô tả dữ liệu

DATASET gives brief description of Tsunamis that were happened between the years 1800 to 2021.

Dataset has 26 columns all including from date metrics to measure metrics with its damages.

Some info:

Tsunami Event Database contains information on the source of the tsunami. It is related to the Tsunami Runup database which contains information on locations where tsunami effects occurred.

Tsunami Runup Database contains information on locations where tsunami effects occurred. It is related to the Tsunami Source Event database which contains information on the source of the tsunami.

Tsunami deposits are the physical evidence left behind when a tsunami impacts a shoreline or affects submarine sediments. These deposits can be preserved in the geologic record and studied long after the tsunami waves dissipate.

tsunami marigrams from analog (microfiche) to high-resolution digital images. Microfiche were created in the 1970s from deteriorating paper records. The collection includes records from 1854 to 1981.

# Khám phá dữ liệu

## Đọc dữ liệu

In [2]:
df = pd.read_csv('Data/tsunami.csv')
df.head()

Year   Mo    Dy    Hr    Mn  Sec  Tsunami Event Validity  \
0     NaN  NaN   NaN   NaN   NaN  NaN                     NaN   
1  1800.0  6.0   2.0  20.0  17.0  NaN                     2.0   
2  1802.0  1.0   4.0   7.0   NaN  NaN                     2.0   
3  1802.0  3.0  19.0   NaN   NaN  NaN                     2.0   
4  1802.0  8.0  10.0   NaN   NaN  NaN                     1.0   

   Tsunami Cause Code  Earthquake Magnitude  Deposits  ...  \
0                 NaN                   NaN       NaN  ...   
1                 1.0                   NaN       0.0  ...   
2                 1.0                   NaN       0.0  ...   
3                 1.0                   NaN       0.0  ...   
4                 0.0                   NaN       0.0  ...   

  Tsunami Magnitude (Iida) Tsunami Intensity  Total Deaths  Total Missing  \
0                      NaN               NaN           NaN            NaN   
1                      NaN               NaN           NaN            NaN   
2                      NaN               3.0           NaN            NaN   
3                      NaN               NaN           NaN            NaN   
4                      NaN               NaN           NaN            NaN   

   Total Missing Description  Total Injuries  Total Damage ($Mil)  \
0                        NaN             NaN                  NaN   
1                        NaN             NaN                  NaN   
2                        NaN             NaN                  NaN   
3                        NaN             NaN                  NaN   
4                        NaN             NaN                  NaN   

   Total Damage Description  Total Houses Destroyed  Total Houses Damaged  
0                       NaN                     NaN                   NaN  
1                       NaN                     NaN                   NaN  
2                       NaN                     NaN                   NaN  
3                       NaN                     NaN                   NaN  
4                       NaN                     NaN                   NaN  

[5 rows x 26 columns]

## Dữ liệu có bao nhiêu dòng và cột ? 

Dựa vào dữ liệu đưa vào ta có thể thấy dữ liệu bị trống ở dòng đầu tiên. Do đó, ta sẽ thực hiệp xóa dòng này trước khi khám phá dữ liệu

In [3]:
df = df.drop(0, axis=0)

In [4]:
n_rows, n_col = df.shape
n_rows, n_col

(2161, 26)

- Dữ liệu có 2161 dòng và 26 cột

## Các dòng dữ liệu

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

In [7]:
df.iloc[:,0:15]

Year    Mo    Dy    Hr    Mn   Sec  Tsunami Event Validity  \
1     1800.0   6.0   2.0  20.0  17.0   NaN                     2.0   
2     1802.0   1.0   4.0   7.0   NaN   NaN                     2.0   
3     1802.0   3.0  19.0   NaN   NaN   NaN                     2.0   
4     1802.0   8.0  10.0   NaN   NaN   NaN                     1.0   
5     1802.0   8.0  15.0   NaN   NaN   NaN                     3.0   
...      ...   ...   ...   ...   ...   ...                     ...   
2157  2021.0   8.0  11.0  17.0  46.0  14.0                     4.0   
2158  2021.0   8.0  12.0  18.0  35.0  20.0                     4.0   
2159  2021.0   8.0  14.0  12.0  29.0   8.0                     4.0   
2160  2021.0   9.0   8.0   1.0  47.0  47.0                     4.0   
2161  2021.0  10.0  12.0   9.0  24.0   3.0                     2.0   

      Tsunami Cause Code  Earthquake Magnitude  Deposits              Country  \
1                    1.0                   NaN       0.0             PORTUGAL   
2                    1.0                   NaN       0.0              CROATIA   
3                    1.0                   NaN       0.0  ANTIGUA AND BARBUDA   
4                    0.0                   NaN       0.0                   UK   
5                    1.0                   NaN       0.0            VENEZUELA   
...                  ...                   ...       ...                  ...   
2157                 1.0                   7.1       0.0          PHILIPPINES   
2158                 1.0                   8.1       0.0         UK TERRITORY   
2159                 1.0                   7.2       0.0                HAITI   
2160                 1.0                   7.0       0.0               MEXICO   
2161                 1.0                   6.4       0.0               GREECE   

                         Location Name  Latitude  Longitude  \
1                               AZORES       NaN        NaN   
2                                BAKAR    45.300     14.400   
3     ANTIGUA ISLAND & ST. CHRISTOPHER    17.200    -62.400   
4                           SW ENGLAND       NaN        NaN   
5                               CUMANA    10.300    -64.500   
...                                ...       ...        ...   
2157      MINDANAO ISLAND, PHILIPPINES     6.455    126.742   
2158           SOUTHERN ATLANTIC OCEAN   -58.416    -25.321   
2159                    PORT-AU-PRINCE    18.408    -73.475   
2160                         S. MEXICO    16.982    -99.773   
2161                             CRETE    35.193     26.256   

      Maximum Water Height (m)  
1                          NaN  
2                          NaN  
3                          NaN  
4                          NaN  
5                          NaN  
...                        ...  
2157                      0.05  
2158                      0.64  
2159                       NaN  
2160                      0.48  
2161                       NaN  

[2161 rows x 15 columns]

- Mỗi dòng ứng với thông tin của 1 cơn sóng thần. Có vẻ không có dòng nào "lạc loài" (nghĩa là không có vấn đề các dòng có ý nghĩa khác nhau).

### Dữ liệu các dòng có bị lặp hay không?

In [6]:
df[df.duplicated(keep=False)]

Year    Mo    Dy    Hr    Mn   Sec  Tsunami Event Validity  \
1449  1961.0  10.0  18.0  16.0  51.0  57.0                     1.0   
1450  1961.0  10.0  18.0  16.0  51.0  57.0                     1.0   

      Tsunami Cause Code  Earthquake Magnitude  Deposits  ...  \
1449                 1.0                   6.5       0.0  ...   
1450                 1.0                   6.5       0.0  ...   

     Tsunami Magnitude (Iida) Tsunami Intensity  Total Deaths  Total Missing  \
1449                      NaN               NaN           NaN            NaN   
1450                      NaN               NaN           NaN            NaN   

      Total Missing Description  Total Injuries  Total Damage ($Mil)  \
1449                        NaN             NaN                  NaN   
1450                        NaN             NaN                  NaN   

      Total Damage Description  Total Houses Destroyed  Total Houses Damaged  
1449                       1.0                     NaN                   NaN  
1450                       1.0                     NaN                   NaN  

[2 rows x 26 columns]

- Vậy dữ liệu ở dòng 1449 và dòng 1950 bị lặp, ta sẽ xóa dòng 1950.

In [7]:
df.drop_duplicates(keep='first',inplace=True)

## Các cột dữ liệu

### Ý nghĩa các cột

In [8]:
df.columns

Index(['Year', 'Mo', 'Dy', 'Hr', 'Mn', 'Sec', 'Tsunami Event Validity',
       'Tsunami Cause Code', 'Earthquake Magnitude', 'Deposits', 'Country',
       'Location Name', 'Latitude', 'Longitude', 'Maximum Water Height (m)',
       'Number of Runups', 'Tsunami Magnitude (Iida)', 'Tsunami Intensity',
       'Total Deaths', 'Total Missing', 'Total Missing Description',
       'Total Injuries', 'Total Damage ($Mil)', 'Total Damage Description',
       'Total Houses Destroyed', 'Total Houses Damaged'],
      dtype='object')

Thông tin các cột có thể đọc thêm tại trang web sau: https://www.ngdc.noaa.gov/hazel/view/hazards/tsunami/event-more-info/2439


|Column|Description|
|----------- |----------- |
|Year|Thời điểm xảy ra sóng thần đơn vị năm.|
|Mo|Thời điểm xảy ra sóng thần đơn vị tháng.|
|Dy|Thời điểm xảy ra sóng thần đơn vị ngày.|
|Hr|Thời điểm xảy ra sóng thần đơn vị giờ.|
|Mn|Thời điểm xảy ra sóng thần đơn vị phút.|
|Sec|Thời điểm xảy ra sóng thần đơn vị giây.|
|Tsunami Event Validity|Độ chuẩn xác của sóng thần **(-1 - 4)** , giá trị càng lớn thì càng chuẩn xác|
|Tsunami Cause Code|Mã nguyên nhân xảy ra sóng thần. **(0 - 11)**|
||1: **động đất**,6: **núi lửa**, 8: **lở đất**
|Earthquake Magnitude|Mức độ tàn phá của động đất (đơn vị Richter).|
|Deposits|Trầm tích sóng thần(trầm tích lắng đọng do kết quả của sóng thần).|
|Country|Đất nước xảy ra sóng thần.|
|Location Name|Địa điểm xảy ra sóng thần (thường là nơi chịu thiệt hại nặng nhất)|
|Latitude|Vĩ độ xảy ra sóng thần.|
||0-90: **bắc bán cầu**; -90-0: **nam bán cầu**
|Longitude|Kinh độ xảy ra sóng thần.|
||0-180: **đông bán cầu**; -180-0: **tây bán cầu**
|Maximum Water Height (m)|Độ cao đạt đỉnh của sóng thần đơn vị mét.|
|Number of Runups|Số địa điểm mà 1 cơn sóng thần tràn vào bờ|
|Tsunami Magnitude (Abe)|Kích cỡ của sóng thần|
|Tsunami Intensity (lida)|Cường độ của sóng thần|
|Total Deaths|Tổng số người chết|
|Total Missing|Tổng số người mất tích|
|Total Missing Description|Mã mô tả tổng số người mất tích **(0-4)**, giá trị càng lớn thiệt hại càng nhiều|
|Total Injuries|Tổng số người bị thương|
|Total Damage ($Mil)|Tổng số thiệt hại (đơn vị 1,000,000 đô la Mỹ)|
|Total Damage Description|Mã mô tả tổng số thiệt hại **(0-4)**, giá trị càng lớn thiệt hại càng nhiều|
|Total Houses Destroyed|Tổng số ngôi nhà bị phá hủy|
|Total Houses Damaged|Tổng số ngôi nhà bị hư hại|

### Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

Dữ liệu ở các cột Tsunami Event Validity,Tsunami Cause Code,Total Missing Description,Total Damage Description sẽ được chuyển từ float thành categorical

In [9]:
# for col in ['Tsunami Event Validity','Tsunami Cause Code','Total Missing Description','Total Damage Description']:
#     df[col] =pd.Categorical(df[col])

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2161 entries, 1 to 2161
Data columns (total 26 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Year                       2161 non-null   float64
 1   Mo                         2112 non-null   float64
 2   Dy                         2054 non-null   float64
 3   Hr                         1338 non-null   float64
 4   Mn                         1289 non-null   float64
 5   Sec                        941 non-null    float64
 6   Tsunami Event Validity     2161 non-null   float64
 7   Tsunami Cause Code         2158 non-null   float64
 8   Earthquake Magnitude       1293 non-null   float64
 9   Deposits                   2161 non-null   float64
 10  Country                    2161 non-null   object 
 11  Location Name              2159 non-null   object 
 12  Latitude                   1892 non-null   float64
 13  Longitude                  1892 non-null   float

Các dữ liệu đều đã ở dạng phù hợp để xử lý

### Với mỗi cột có kiểu dữ liệu dạng số (numerical), các giá trị được phân bố như thế nào

Với các cột có kiểu dữ liệu số, ta sẽ tính:
- Tỉ lệ % (từ 0 đến 100) các giá trị thiếu 
- Giá trị min
- Giá trị lower quartile (phân vị 25)
- Giá trị median (phân vị 50)
- Giá trị upper quartile (phân vị 75)
- Giá trị max

In [11]:
def calculate_quartile(data):
    nume_col_info_df = pd.DataFrame()
    for col in data.keys():
        if data[col].dtype in [np.int64, np.float64] :
            missing = data[col].isnull().sum()
            missing_percentage = round(data[col].isnull().sum() * 100 / len(data[col]), 1)
            min_value = data[col].min()
            lower_quartile = data[col].quantile(0.25)
            median = data[col].median()
            upper = data[col].quantile(0.75)
            max_value = data[col].max()
            row_line = pd.Series([missing,missing_percentage, min_value, lower_quartile, median, upper, max_value], 
                                 index = ['num_missing','missing_percentage', 'min', 'lower_quartile', 'median', 'upper_quartile', 'max'])
            nume_col_info_df[col] = row_line
    return nume_col_info_df

In [12]:
info_df = calculate_quartile(df)
info_df

Year    Mo     Dy     Hr     Mn     Sec  \
num_missing            0.0  49.0  107.0  823.0  872.0  1220.0   
missing_percentage     0.0   2.3    5.0   38.1   40.4    56.5   
min                 1800.0   1.0    1.0    0.0    0.0     0.0   
lower_quartile      1877.0   4.0    8.0    5.0   14.0    10.8   
median              1927.0   7.0   16.0   11.0   29.0    25.7   
upper_quartile      1977.0  10.0   23.0   18.0   43.0    42.0   
max                 2021.0  12.0   31.0   23.0   59.0    59.9   

                    Earthquake Magnitude  Deposits  Latitude  Longitude  \
num_missing                        868.0       0.0   269.000   269.0000   
missing_percentage                  40.2       0.0    12.500    12.5000   
min                                  3.7       0.0   -63.000  -179.9710   
lower_quartile                       6.5       0.0    -8.300   -71.4585   
median                               7.0       0.0    16.000    51.9400   
upper_quartile                       7.6       0.0    38.393   140.4565   
max                                  9.5     144.0    71.813   179.9840   

                    Maximum Water Height (m)  Number of Runups  \
num_missing                          1051.00               0.0   
missing_percentage                     48.70               0.0   
min                                     0.01               0.0   
lower_quartile                          0.20               1.0   
median                                  1.00               1.0   
upper_quartile                          3.50               3.0   
max                                   524.60            6288.0   

                    Tsunami Magnitude (Iida)  Tsunami Intensity  Total Deaths  \
num_missing                          1584.00            1263.00       1668.00   
missing_percentage                     73.30              58.50         77.20   
min                                    -4.64              -4.14          1.00   
lower_quartile                         -1.00               0.00          4.00   
median                                  0.70               1.00         24.50   
upper_quartile                          2.00               2.00        262.75   
max                                     8.40               9.00     316000.00   

                    Total Missing  Total Injuries  Total Damage ($Mil)  \
num_missing                2149.0          1879.0            2017.0000   
missing_percentage           99.5            87.0              93.4000   
min                           1.0             1.0               0.0030   
lower_quartile               20.0            16.0               2.4125   
median                      100.0           100.0              39.0000   
upper_quartile              498.0           500.0             520.0000   
max                       43476.0        374171.0          220136.6000   

                    Total Houses Destroyed  Total Houses Damaged  
num_missing                         1963.0                2083.0  
missing_percentage                    90.9                  96.4  
min                                    1.0                   2.0  
lower_quartile                        50.0                 127.0  
median                               588.0                1000.0  
upper_quartile                      5043.0               10044.0  
max                              5360000.0              500000.0

#### Số-lượng/tỉ-lệ các giá trị thiếu?

In [13]:
pd.DataFrame([info_df.loc['num_missing'],info_df.loc['missing_percentage']]).transpose()

num_missing  missing_percentage
Year                              0.0                 0.0
Mo                               49.0                 2.3
Dy                              107.0                 5.0
Hr                              823.0                38.1
Mn                              872.0                40.4
Sec                            1220.0                56.5
Earthquake Magnitude            868.0                40.2
Deposits                          0.0                 0.0
Latitude                        269.0                12.5
Longitude                       269.0                12.5
Maximum Water Height (m)       1051.0                48.7
Number of Runups                  0.0                 0.0
Tsunami Magnitude (Iida)       1584.0                73.3
Tsunami Intensity              1263.0                58.5
Total Deaths                   1668.0                77.2
Total Missing                  2149.0                99.5
Total Injuries                 1879.0                87.0
Total Damage ($Mil)            2017.0                93.4
Total Houses Destroyed         1963.0                90.9
Total Houses Damaged           2083.0                96.4

#### Min? max? → Có gì bất thường không?

In [14]:
pd.DataFrame([info_df.loc['min'],info_df.loc['max']]).transpose()

min          max
Year                      1800.000     2021.000
Mo                           1.000       12.000
Dy                           1.000       31.000
Hr                           0.000       23.000
Mn                           0.000       59.000
Sec                          0.000       59.900
Earthquake Magnitude         3.700        9.500
Deposits                     0.000      144.000
Latitude                   -63.000       71.813
Longitude                 -179.971      179.984
Maximum Water Height (m)     0.010      524.600
Number of Runups             0.000     6288.000
Tsunami Magnitude (Iida)    -4.640        8.400
Tsunami Intensity           -4.140        9.000
Total Deaths                 1.000   316000.000
Total Missing                1.000    43476.000
Total Injuries               1.000   374171.000
Total Damage ($Mil)          0.003   220136.600
Total Houses Destroyed       1.000  5360000.000
Total Houses Damaged         2.000   500000.000

### Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical), các giá trị được phân bố như thế nào?

Với các cột có kiểu dữ liệu không phải số, ta sẽ tính:

- Tỉ lệ % (từ 0 đến 100) các giá trị thiếu
- Số lượng các giá trị (các giá trị ở đây là các giá trị khác nhau và ta không xét giá trị thiếu) 

- Tỉ lệ % (từ 0 đến 100) của mỗi giá trị được sort theo tỉ lệ % giảm dần (ta không xét giá trị thiếu, tỉ lệ là tỉ lệ so với số lượng các giá trị không thiếu): bạn dùng dictionary để lưu, key là giá trị, value là tỉ lệ %; với cột mà ứng với câu hỏi dạng multichoice, cách làm tương tự như ở trên

**Lưu ý, ta sẽ đưa cả các cột có dạng số nhưng có ít hơn 15 giá trị khác nhau vào dạng categorical**

In [28]:
unique_counts = df.nunique()
lowest_distinct = unique_counts < 15
lowest_distinct = unique_counts[lowest_distinct == True]
lowest_distinct

Mo                           12
Tsunami Event Validity        6
Tsunami Cause Code           12
Total Missing                10
Total Missing Description     4
Total Damage Description      4
dtype: int64

Trừ cột **Total Missing** (vì cột này thống kế tổng số người mất tích sau mỗi trận sóng thần nên không thể là categorical) thì các cột còn lại đều có thể xét vào dạng categorical

In [30]:
cat = lowest_distinct.index

In [31]:
#cat = df.select_dtypes(exclude=["number","bool_"]).columns
missing = [];missing_percentage=[];num_values=[];value_percentages=[]

for column in cat:
    missing.append(df[column].isnull().sum())
    missing_percentage.append((df[column].isnull().sum() * 100 / len(df)).round(1))
    temp = df[column].dropna()
    num_values.append(len(temp.unique()))
    value_percentages.append(((temp.value_counts(normalize=True)*100).round(1)).to_dict())
cat_info_df = pd.DataFrame([missing,missing_percentage,num_values,value_percentages],
                            index=['num_missing','missing_percentage','num_values','value_percentages'],
                            columns=list(cat))
cat_info_df

Mo  \
num_missing                                                        49   
missing_percentage                                                2.3   
num_values                                                         12   
value_percentages   {8.0: 10.6, 11.0: 9.6, 1.0: 9.0, 5.0: 8.8, 3.0...   

                                               Tsunami Event Validity  \
num_missing                                                         0   
missing_percentage                                                  0   
num_values                                                          6   
value_percentages   {4.0: 38.5, 2.0: 19.2, 1.0: 16.7, 3.0: 15.3, -...   

                                                   Tsunami Cause Code  \
num_missing                                                         3   
missing_percentage                                                0.1   
num_values                                                         12   
value_percentages   {1.0: 70.5, 0.0: 9.7, 8.0: 4.8, 3.0: 4.4, 6.0:...   

                                                        Total Missing  \
num_missing                                                      2150   
missing_percentage                                               99.5   
num_values                                                         10   
value_percentages   {30.0: 18.2, 43476.0: 9.1, 329.0: 9.1, 10.0: 9...   

                                     Total Missing Description  \
num_missing                                               2148   
missing_percentage                                        99.4   
num_values                                                   4   
value_percentages   {1.0: 46.2, 3.0: 38.5, 4.0: 7.7, 2.0: 7.7}   

                                        Total Damage Description  
num_missing                                                 1254  
missing_percentage                                            58  
num_values                                                     4  
value_percentages   {1.0: 38.6, 2.0: 27.9, 3.0: 19.1, 4.0: 14.4}

In [32]:
df['Mo'].value_counts()

8.0     223
11.0    202
1.0     191
5.0     185
3.0     181
10.0    180
9.0     176
12.0    167
2.0     155
4.0     155
7.0     151
6.0     146
Name: Mo, dtype: int64

#### Số-lượng/tỉ-lệ các giá trị thiếu?

In [16]:
pd.DataFrame([cat_info_df.loc['num_missing'],cat_info_df.loc['missing_percentage']]).transpose()

num_missing  missing_percentage
Tsunami Event Validity             0.0                 0.0
Tsunami Cause Code                 3.0                 0.1
Country                            0.0                 0.0
Location Name                      2.0                 0.1
Total Missing Description       2147.0                99.4
Total Damage Description        1254.0                58.1

#### Số lượng các giá trị khác nhau? Show một vài giá trị → Có gì bất thường không?

In [17]:
country_df = pd.DataFrame([cat_info_df['Country']['value_percentages']])
country_df.T.head(10)

0
USA               12.0
INDONESIA         10.6
JAPAN             10.4
CHILE              6.9
GREECE             5.2
ITALY              4.2
RUSSIA             4.2
NEW ZEALAND        4.0
PHILIPPINES        3.7
PAPUA NEW GUINEA   3.5

In [18]:
pd.DataFrame([cat_info_df['Location Name']['value_percentages']]).T.head(10)

0
NORTHERN CHILE    3.3
CENTRAL CHILE     2.4
HAWAII            1.9
SANRIKU           1.7
S. KURIL ISLANDS  1.6
S. MEXICO         1.5
N. CALIFORNIA     1.4
BISMARCK SEA      1.3
SOLOMON ISLANDS   1.3
BANDA SEA         1.2

In [19]:
pd.DataFrame([cat_info_df['Tsunami Cause Code']['value_percentages']]).T.head(10)

0
1.0   70.5
0.0    9.7
8.0    4.8
3.0    4.5
6.0    4.2
9.0    4.2
2.0    0.7
4.0    0.6
7.0    0.5
11.0   0.1

- Có vẻ dữ liệu đều ổn không có gì bất thường xảy ra ở đây

# Đưa ra các câu hỏi có ý nghĩa cần trả lời

## Câu hỏi - Lợi ích

# Tiền xử lý + phân tích dữ liệu để trả lời cho từng câu hỏi

# Nhìn lại quá trình làm đồ án

## Những khó khăn

## Học được gì hữu ích

## Nếu có thêm thời gian sẽ làm gì

# Tài liệu tham khảo